multinet/workspace.py

```diff
@@ -0,0 +1,356 @@
+from arango.exceptions import DatabaseCreateError, EdgeDefinitionCreateError
+from arango.cursor import Cursor
+
+from multinet import util
+from multinet.types import EdgeTableProperties
+from multinet.validation import ValidationFailure, UndefinedTable, UndefinedKeys
+from multinet.validation.csv import validate_csv
+from multinet.db import (
+    workspace_mapping,
+    workspace_mapping_collection,
+    db,
+    system_db,
+    _run_aql_query,
+)
+from multinet.errors import (
+    AlreadyExists,
+    ValidationFailed,
+    InternalServerError,
+    WorkspaceNotFound,
+    BadQueryArgument,
+    GraphNotFound,
+    TableNotFound,
+    GraphCreationError,
+)
+from multinet.user import User
+from multinet.graph import Graph
+from multinet.table import Table
+
+from typing import Any, List, Dict, Generator, Optional
+
+
+@dataclass
+class WorkspacePermissions:
+    """The permissions on a workspace."""
+
+    # TODO: Change str to User once updating permissions storage
+    owner: str
+    maintainers: List[str] = dataclasses.field(default_factory=lambda: [])
+    writers: List[str] = dataclasses.field(default_factory=lambda: [])
+    readers: List[str] = dataclasses.field(default_factory=lambda: [])
+    public: bool = False
+
+
+class Workspace:
+    """Workspace."""
+
+    # Keys that aren't saved to the database
+    exclude_keys = {"handle", "readonly_handle", "readonly"}
+
+    def __init__(self, name: str):
+        """Create a Workspace Object."""
+        self.name = name
+
+        # TODO: Don't access database right away
+        doc = self.get_metadata()
+
+        # Due to call above, doc is guaranteed to be valid
+        self.internal: str = doc["internal"]
+        self.permissions: WorkspacePermissions = WorkspacePermissions(
+            **doc["permissions"]
+        )
+
+        self.readonly_handle = db(self.internal)
+        self.handle = db(self.internal, readonly=False)
+
+    @staticmethod
+    def exists(name: str) -> bool:
+        """Return if this workspace exists or not."""
+        return bool(workspace_mapping(name))
+
+    @staticmethod
+    def create(name: str, owner: User) -> Workspace:
+        """Create a workspace, owned by `owner`."""
+        if Workspace.exists(name):
+            raise AlreadyExists("Workspace", name)
+
+        internal = util.generate_arango_workspace_name()
+
+        try:
+            system_db(readonly=False).create_database(internal)
+        except DatabaseCreateError:
+            # Could only happen if there's a name collision
+            raise InternalServerError("Error creating workspace")
+
+        permissions = WorkspacePermissions(owner=owner.sub).__dict__
+        workspace_dict = {
+            "name": name,
+            "internal": internal,
+            "permissions": permissions,
+        }
+
+        coll = workspace_mapping_collection(readonly=False)
+        coll.insert(workspace_dict, sync=True)
+        workspace_mapping.cache_clear()
```

Comment:
We could, but I didn't remove it here because I think if we're going to be possibly impacting the performance of the application, it makes sense to do that in it's own PR, so that the effects are easier to observe.

Reply:
Related to #453

Reply:
Great. I forgot we were talking about using `memcached`!